pkg/s3select/progress.go

```diff
@@ -0,0 +1,90 @@
+/*
+ * Minio Cloud Storage, (C) 2019 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package s3select
+
+import (
+	"compress/bzip2"
+	"fmt"
+	"io"
+	"sync/atomic"
+
+	gzip "github.com/klauspost/pgzip"
+)
+
+type countUpReader struct {
+	reader    io.Reader
+	bytesRead int64
+}
+
+func (r *countUpReader) Read(p []byte) (n int, err error) {
+	n, err = r.reader.Read(p)
+	atomic.AddInt64(&r.bytesRead, int64(n))
```

Comment:
Does this need to be atomic? In Parquet input, is data being read concurrently?

Reply:
I think it is being read concurrently.

Reply:
* When parquet is concerned, it requires random access to an input source.  However minio internal or any http response does not provide random accessible stream.
* This is fixed by having a callback function `func(offset, length int64) (io.ReadCloser, error)`.  Whenever parquet data is read data in different position/length, it uses callback function to get new `io.ReadCloser` and performs decoding.
* Parquet comes with a limitation when s3select is concerned. Its not possible to provide `BytesRead` and `BytesScanned` in progress and stats messages (because of above nature of reading).  Hence `-1` is returned for both, , but `BytesReturned` is sent.
* `countUpReader` is not used for parquet.